### PR TITLE
Update TO_DATE, TO_TIMESTAMP scalar functions to support LargeUtf8, Utf8View

### LINKED ISSUE #12928: Update TO_DATE, TO_TIMESTAMP scalar functions to support LargeUtf8, Utf8View
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/datafusion/issues/11752 and https://github.com/apache/datafusion/issues/11790

Currently, a call to TO_DATE or TO_TIMESTAMP* UDFs with a Utf8View datatypes fails. After the change that fixes this issue, it should not.

```
> create table ts_utf8_data(ts varchar(100), format varchar(100)) as values
  ('2020-09-08 12/00/00+00:00', '%Y-%m-%d %H/%M/%S%#z'),
  ('2031-01-19T23:33:25+05:00', '%+'),
  ('08-09-2020 12:00:00+00:00', '%d-%m-%Y %H:%M:%S%#z'),
  ('1926632005', '%s'),
  ('2000-01-01T01:01:01+07:00', '%+');
0 row(s) fetched. 
Elapsed 0.062 seconds.

> create table ts_utf8view_data as
select arrow_cast(ts, 'Utf8View') as ts, arrow_cast(format, 'Utf8View') as format from ts_utf8_data;
0 row(s) fetched. 
Elapsed 0.010 seconds.

> SELECT to_timestamp(t.ts, t.format),
       to_timestamp_seconds(t.ts, t.format),
       to_timestamp_millis(t.ts, t.format),
       to_timestamp_micros(t.ts, t.format),
       to_timestamp_nanos(t.ts, t.format)
       from ts_utf8view_data as t;
Execution error: to_timestamp function unsupported data type at index 1: Utf8View
```

We are working to add complete StringView support in DataFusion, which permits potentially much faster processing of string data. See https://github.com/apache/datafusion/issues/10918 for more background.

Today, most DataFusion string functions support DataType::Utf8 and DataType::LargeUtf8 and when called with a StringView argument DataFusion will cast the argument back to DataType::Utf8 which is expensive.

To realize the full speed of StringView, we need to ensure that all string functions support the DataType::Utf8View directly.

### Describe the solution you'd like

Update the functions to support DataType::Utf8View directly

### Describe alternatives you've considered

_No response_

### Additional context

The typical steps are:

    Write some tests showing the function doesn't support Utf8View (see the tests in [string_view.slt](https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/string_view.slt) to ensure the arguments are not being cast
    Change the Signature of the function to accept Utf8View in addition to Utf8/LargeUtf8
    Update the implementation of the function to operate on Utf8View

Example PRs

    Update to use an arrow kernel that already supports StringView: 

https://github.com/apache/datafusion/pull/11787
Change the implementation to support StringView directly:
https://github.com/apache/datafusion/pull/11676
Change implementation (option 2):
https://github.com/apache/datafusion/pull/11556

### PR DESCRIPTION
**Pull request by Omega359**
## Which issue does this PR close?

Closes #12928

## Rationale for this change

Update functions to support Utf8View and LargeUtf8

## What changes are included in this PR?

Code, tests, benchmark

## Are these changes tested?

Yes

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/functions/src/datetime/common.rs [hunk 1]
```diff
@@ -178,63 +180,80 @@ pub(crate) fn string_to_timestamp_millis_formatted(s: &str, format: &str) -> Res
...
+                let result = a.as_ref().map(|x| op(x)).transpose()?;
                 Ok(ColumnarValue::Scalar(S::scalar(result)))
             }
             other => exec_err!("Unsupported data type {other:?} for function {name}"),
         },
     }
 }
 
-// given an function that maps a `&str`, `&str` to an arrow native type,
+// given a function that maps a `&str`, `&str` to an arrow native type,
```
Comment (findepi) [inline datafusion/functions/src/datetime/common.rs:227]: While at it, can we uppercase "Given"?

File: datafusion/sqllogictest/test_files/timestamps.slt [hunk 2]
```diff
@@ -2201,23 +2209,102 @@ SELECT to_timestamp(t.ts, t.format) from ts_utf8_data as t
...
+2031-01-19T23:33:25 2031-01-19T23:33:25 2031-01-19T23:33:25 2031-01-19T23:33:25 2031-01-19T23:33:25
+1999-12-31T18:01:01 1999-12-31T18:01:01 1999-12-31T18:01:01 1999-12-31T18:01:01 1999-12-31T18:01:01
+
+# verify timestamp data using tables with formatting options where at least one column cannot be parsed
+query error Error parsing timestamp from '1926632005' using format '%d-%m-%Y %H:%M:%S%#z': input contains invalid characters
+SELECT to_timestamp(t.ts, '%Y-%m-%d %H/%M/%S%#z', '%+', '%d-%m-%Y %H:%M:%S%#z') from ts_utf8_data as t
+
+# verify timestamp data using tables with formatting options where one of the formats is invalid
 query P
-SELECT to_timestamp(t.ts, '%Y-%m-%d %H/%M/%S%#z', '%+', '%s', '%d-%m-%Y %H:%M:%S%#z') from ts_utf8_data as t
```
Comment (findepi) [inline datafusion/sqllogictest/test_files/timestamps.slt:2206]: Where did it go to?

**Review by findepi — APPROVED** [review]

File: datafusion/sqllogictest/test_files/timestamps.slt (hunk 2 above)
Comment (Omega359) [reply datafusion/sqllogictest/test_files/timestamps.slt:2206]: hmm. Incorrect refactor ... I changed the sql a few times here to condense the tests and it looks like I may have inadvertently changed this.

Comment (Omega359) [reply datafusion/sqllogictest/test_files/timestamps.slt:2206]: Actually, it's now part of the sql starting at line 2242

File: datafusion/functions/benches/to_timestamp.rs [hunk 3]
```diff
@@ -51,67 +147,66 @@ fn criterion_benchmark(c: &mut Criterion) {
...
-        inputs.append_value("1997-01-31 092656+04:00");
-        format1_builder.append_value("%+");
-        format2_builder.append_value("%c");
-        format3_builder.append_value("%Y-%m-%d %H%M%S%:z");
-
-        inputs.append_value("Sun Jul  8 00:34:60 2001");
-        format1_builder.append_value("%+");
-        format2_builder.append_value("%c");
-        format3_builder.append_value("%Y-%m-%d 00:00:00");
+    c.bench_function("to_timestamp_with_formats_utf8", |b| {
```
Comment (alamb) [inline datafusion/functions/benches/to_timestamp.rs:150]: 👍

File: datafusion/functions/src/datetime/common.rs (hunk 1 above)
Comment (alamb) [reply datafusion/functions/src/datetime/common.rs:227]:
```suggestion
// Given a function that maps a `&str`, `&str` to an arrow native type,
```

**Review by alamb — APPROVED** [review]
Thank you @Omega359  and @findepi  -- this is looking nice now

**Comment by alamb** [comment]
I merged up from main and applied @findepi 's comment suggestion. If the tests pass I think this one is good to go

**Comment by alamb** [comment]
🚀

**Pull request merged**